site/docs/design-docs/018-kinesis-connectors.md

```diff
@@ -0,0 +1,476 @@
+  sized stream, as far as shard count is concerned - ultimately the
+  owner of the stream is responsible for setting up enough shards to be
+  able to handle his data rates)
+
+The flow control process is _adaptive_ in the sense that:
+
+* it kicks in only when batches start failing due to shard ingestion
+  rates being tripped
+* until failures repeat, it keeps quickly increasing the sleep delays
+  to stop them from happening
+* once failures stop, it slowly decreases the sleep delays until they
+  are eliminated (i.e., the data volume spike was only temporary) or
+  until failures start happening again
+
+Under normal circumstances, if there are enough shards in the stream and
+their data ingestion rate covers the data flow, this whole flow control
+process stays shut off. The sink publishes data with the lowest possible
+latency.
+
+### Discovery
+
+As we've seen in the [flow control](#flow-control) section, one element
+used to control the throughput is batch size. Under normal conditions,
+the sink uses the default/maximum batch size of 500. When flow control
+kicks a new batch size is picked as a function of the number of open
+shards in the stream.
+
+For this to happen, the sinks need to have relatively up-to-date
+information about the count of open shards. The sink achieves this by
+using a mechanism very similar to the [discovery process employed by the
+source](#discovery). The only real difference is that the sinks use the
+[DescribeStreamSummary](https://docs.aws.amazon.com/kinesis/latest/APIReference/API_DescribeStreamSummary.html)
+operation instead of the
+[ListShards](https://docs.aws.amazon.com/kinesis/latest/APIReference/API_ListShards.html)
+one.
+
+### Write Order
+
+Under normal circumstances, the Kinesis sink preserves the order of
+items belonging to the same partition key. However, when the [flow
+control](#flow-control) mechanism kicks in, the ordering might be lost
+on occasion.
+
+This unfortunate fact originates in the way how KDS handles shard
+ingestion rate violations. When KDS receives a batch to be ingested, it
+processes each item in it one-by-one, and if some fail, it doesn't stop
+processing the batch. The result is that some items from a batch get
+rejected, some get ingested, but in a random manner. The sink does
+resend the un-ingested item, they won't get lost, but there is nothing
+it can do to preserve the initial ordering. This is very unfortunate
+since all that would be needed for avoiding this problem would be for
+Kinesis to reject all remaining items from the batch once one of them
+trips the ingestion rate limit. The designers probably made the choice
+they did to save on data traffic.
+
+Nothing that we can do about it at this point needs to be documented,
+and users of the sink need to be aware.
+
+### Fault Tolerance
+
+Since there is no transaction support in Kinesis the sink can't support
+_exactly-once_ processing. It can, however, support _at-least-once_
```

Comment:
Since it's a sink, from its point of view isn't delivery and processing the same thing?

Reply:
I think "processing" should apply to the whole pipeline. The main problem with at-least-once processing in Jet is that the aggregation results are just wrong. If the only bad outcome is getting the exact same, correct data twice in the output (properly timestamped, so pertaining to the same window), it would still count as "exactly-once processing".